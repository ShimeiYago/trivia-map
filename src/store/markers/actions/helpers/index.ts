import { Marker } from '../../model';

/**
 * Concats two marker list.
 *
 * @param origMarkers - The original marker list
 * @param newMarkers - The marker list to add newly
 * @returns Combined marker list
 */
export function concatMarkers(
  origMarkers: Marker[],
  newMarkers: Marker[],
): Marker[] {
  return [...origMarkers, ...newMarkers];
}

/**
 * Add new marker to existing marker list.
 * If a marker with same markerId exists in list, numberOfPublicArticles of the marker is increased.
 *
 * @param origMarkers - The original marker list
 * @param newMarker - The marker to add newly
 * @returns updated marker list
 */
export function pushMarker(origMarkers: Marker[], newMarker: Marker): Marker[] {
  const editableOrigMarker = deepCopyMarkers(origMarkers);

  const sameMarkerIndex = editableOrigMarker.findIndex(
    (marker) => marker.markerId === newMarker.markerId,
  );

  if (sameMarkerIndex === -1) {
    editableOrigMarker.push(newMarker);
  } else {
    editableOrigMarker[sameMarkerIndex].numberOfPublicArticles += 1;
  }
  return editableOrigMarker;
}

/**
 * Delete a marker from existing marker list.
 * If target marker have multiple articles, numberOfPublicArticles of the marker is decreaded.
 *
 * @param origMarkers - The original marker list
 * @param deletingMarkerId - The marker id to delete
 * @returns updated marker list
 */
export function deleteOneMarker(
  origMarkers: Marker[],
  deletingMarkerId: number,
): Marker[] {
  const editableOrigMarker = deepCopyMarkers(origMarkers);

  const sameMarkerIndex = editableOrigMarker.findIndex(
    (marker) => marker.markerId === deletingMarkerId,
  );

  if (sameMarkerIndex === -1) {
    return editableOrigMarker;
  }

  if (editableOrigMarker[sameMarkerIndex].numberOfPublicArticles > 1) {
    editableOrigMarker[sameMarkerIndex].numberOfPublicArticles -= 1;
  } else {
    editableOrigMarker.splice(sameMarkerIndex, 1);
  }
  return editableOrigMarker;
}

function deepCopyMarkers(markers: Marker[]) {
  return markers.map((marker) => {
    return {
      markerId: marker.markerId,
      lat: marker.lat,
      lng: marker.lng,
      park: marker.park,
      numberOfPublicArticles: marker.numberOfPublicArticles,
    };
  });
}
