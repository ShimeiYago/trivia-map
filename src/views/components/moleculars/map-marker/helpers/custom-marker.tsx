// TODO: This component should be completed without using any type.

import React, { useEffect, useRef, useState } from 'react';
import { Marker, Popup } from 'react-leaflet';
import {
  LatLng,
  LeafletEventHandlerFnMap,
  Map as LeafletMap,
  Marker as MarkerType,
  Icon,
} from 'leaflet';

export const CustomMarker = React.forwardRef((props: Props, ref) => {
  const { position, icon, draggable, eventHandlers, popup, autoOpen, map } =
    props;

  const [refReady, setRefReady] = useState(false);
  let popupRef = useRef();

  useEffect(() => {
    if (refReady && autoOpen) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (map as any).openPopup(popupRef, position);
    }
  }, [autoOpen, refReady, map]);

  return (
    <Marker
      position={position}
      icon={icon}
      draggable={draggable}
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ref={ref as any}
      eventHandlers={eventHandlers}
    >
      <Popup
        ref={(r) => {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          popupRef = r as any;
          setRefReady(true);
        }}
      >
        {popup}
      </Popup>
    </Marker>
  );
});

type Props = {
  position: LatLng;
  icon: Icon;
  popup: React.ReactNode;
  autoOpen: boolean;
  map: LeafletMap;
  draggable?: boolean;
  ref: React.RefObject<MarkerType>;
  eventHandlers: LeafletEventHandlerFnMap;
};
