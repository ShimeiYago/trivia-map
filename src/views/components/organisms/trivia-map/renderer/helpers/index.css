.marker-cluster-red {
  background-color: rgba(218, 94, 94, 0.6);
  }
.marker-cluster-red div {
  background-color: rgba(226, 36, 36, 0.6);
  color: white;
  font-size: 14px;
}