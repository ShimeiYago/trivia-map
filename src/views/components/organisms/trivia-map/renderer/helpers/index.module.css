.popup-content {
  width: 250px;
}
