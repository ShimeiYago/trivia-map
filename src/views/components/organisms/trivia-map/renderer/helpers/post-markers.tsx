import React from 'react';
import { Typography } from '@mui/material';
import { MarkerDict } from 'store/markers/model';
import { LatLng, Map as LeafletMap } from 'leaflet';
import { MapMarker } from 'views/components/moleculars/map-marker';

export class PostMarkers extends React.Component<Props> {
  render() {
    const { map, markers, popupDisabled, hiddenMarkerIds } = this.props;

    const markerElmList = Object.keys(markers).map((postId) => {
      if (hiddenMarkerIds.includes(postId)) {
        return null;
      }

      const marker = markers[postId];

      const popup = !popupDisabled && this.renderPopupContents(marker.title);

      return (
        <MapMarker
          map={map}
          position={new LatLng(marker.position.lat, marker.position.lng)}
          popup={popup}
          key={`post-marker-${postId}`}
        />
      );
    });

    return <>{markerElmList}</>;
  }

  protected renderPopupContents = (title: string) => (
    <Typography>{title}</Typography>
  );
}

export type Props = {
  map: LeafletMap;
  markers: MarkerDict;
  popupDisabled: boolean;
  hiddenMarkerIds: string[];
};
