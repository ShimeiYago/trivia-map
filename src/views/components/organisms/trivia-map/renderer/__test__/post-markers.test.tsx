import { shallow, ShallowWrapper } from 'enzyme';
import { Map as LeafletMap, MarkerCluster } from 'leaflet';
import { PostMarkers, Props } from '../helpers/post-markers';

const basicProps: Props = {
  markers: [
    {
      markerId: 100,
      lat: 0,
      lng: 0,
      park: 'S',
      numberOfPublicArticles: { total: 1, eachCategory: [1, 0, 0] },
    },
    {
      markerId: 101,
      lat: 0,
      lng: 0,
      park: 'S',
      numberOfPublicArticles: { total: 4, eachCategory: [1, 2, 1] },
    },
  ],
  hiddenMarkerIds: [],
  map: { setView: jest.fn() } as unknown as LeafletMap,
  popupDisabled: false,
  openFormWithTheMarker: jest.fn(),
  editting: false,
  isMobile: false,
  showNumberOfArticles: true,
};

let shallowWrapper: ShallowWrapper<Props, unknown, PostMarkers>;

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    shallowWrapper = shallow(<PostMarkers {...basicProps} />);
  });

  it('basic', () => {
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('with hiddenMarkerIds', () => {
    shallowWrapper.setProps({ hiddenMarkerIds: [100] });
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('form is editting', () => {
    shallowWrapper.setProps({ editting: true });
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('category is selected', () => {
    shallowWrapper.setProps({ categoryId: 1 });
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('without number of articles', () => {
    shallowWrapper.setProps({ showNumberOfArticles: false });
    expect(shallowWrapper).toMatchSnapshot();
  });
});

describe('handleClickAdd', () => {
  it('should call openFormWithTheMarker', () => {
    shallowWrapper = shallow(<PostMarkers {...basicProps} />);
    const instance = shallowWrapper.instance();

    instance['handleClickAdd']({
      lat: 1,
      lng: 1,
      park: 'S',
    })();
    expect(instance.props.openFormWithTheMarker).toBeCalled();
  });
});

describe('clusterIconCreator', () => {
  it('should return div icon', () => {
    shallowWrapper = shallow(<PostMarkers {...basicProps} />);
    const instance = shallowWrapper.instance();

    const markerCluster = {
      getChildCount: () => 2,
    } as unknown as MarkerCluster;

    const divIcon = instance['clusterIconCreator'](markerCluster);
    expect(divIcon).toBeDefined();
  });
});
