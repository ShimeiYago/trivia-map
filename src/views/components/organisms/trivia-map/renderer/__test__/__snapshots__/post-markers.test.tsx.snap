// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<Fragment>
  <MapMarker
    autoOpen={false}
    draggable={false}
    key="post-marker-001"
    map={
      Object {
        "setView": [MockFunction],
      }
    }
    popup={
      <ForwardRef(Typography)>
        title2
      </ForwardRef(Typography)>
    }
    position={
      Object {
        "lat": 0,
        "lng": 0,
      }
    }
    variant="blue"
  />
</Fragment>
`;
